slyd/app/utils/ferry-websocket.js

```diff
@@ -0,0 +1,215 @@
+import Ember from 'ember';
+import config from '../config/environment';
+import ApplicationUtils from '../mixins/application-utils';
+
+/* global URI */
+
+const DEFAULT_RECONNECT_TIMEOUT = 5000;
+const DEFAULT_MAX_RECONNECT_TIMEOUT = 60000;
+const DEFAULT_COMMANDS = Object.freeze({
+    heartbeat: function() {},
+    saveChanges: function() {},
+    delete: function() {},
+    rename: function() {},
+    resolve: function() {},
+});
+
+var defaultUrl = function() {
+    var uri = URI.parse(config.SLYD_URL || window.location.protocol + '//' + window.location.host);
+    if (!/wss?/.test(uri.protocol)) {
+        uri.protocol = uri.protocol === 'https' ? 'wss' : 'ws';
+    }
+    uri.path = '/ws';
+    return URI.build(uri);
+};
+
+export default Ember.Object.extend(ApplicationUtils, {
+    init: function(options) {
+        options = options || {};
+        this.set('commands', options.commands || {});
+        this.set('url', options.url || defaultUrl());
+        this.set('cleanup', options.cleanup || function() {});
+        this.set('init', options.init || function() {});
+        this.set('protocols', options.protocols);
+        this.set('commands', this._buildCommands(options.commands || {}));
+        this.set('ws', null);
+        this.set('reconnectTimeout', DEFAULT_RECONNECT_TIMEOUT);
+        this.set('closed', true);
+        this.set('opened', false);
+        this.set('connecting', false);
+        this.set('nextConnect', null);
+        this.set('deferreds', {});
+        this.heartbeat = null;
+    },
+
+    connect: function() {
+        return this._createWebsocket();
+    },
+
+    _createWebsocket: function() {
+        var ws, deferred = new Ember.RSVP.defer();
+        if (this.get('nextConnect')) {
+            Ember.run.cancel(this.get('nextConnect'));
+            this.set('nextConnect', null);
+        }
+        try {
+            this.set('connecting', true);
+            this.notifyPropertyChange('connecting');
+            ws = new WebSocket(this.get('url'));
+        } catch (err) {
+            Ember.Logger.log('Error connecting to server: ' + err);
+        } finally {
+            this.set('connecting', false);
+            this.notifyPropertyChange('connecting');
+        }
```

Comment:
The error handling isn't perfect, in the catch clause reject the deferred and return. Should we wait until `onopen` to set `connecting` back to `false`?